File: src/main/java/com/derongan/minecraft/madeinabyss/Ascension/Effect/Effects/AbstractAscensionEffect.java
package com.derongan.minecraft.madeinabyss.Ascension.Effect.Effects;

import com.derongan.minecraft.madeinabyss.AbyssContext;
import org.bukkit.entity.Player;

public abstract class AbstractAscensionEffect implements AscensionEffect {
    private AbyssContext context;
    int durationRemaining;
    int elapsed;
    int strength;
    private long offset; //TODO honor this value

    AbstractAscensionEffect(AbyssContext context, long offset, int strength, int durationRemaining) {
        this.context = context;
        this.durationRemaining = durationRemaining;
        this.elapsed = 0;
        this.offset = offset;
        this.strength = strength;
    }

    AbyssContext getContext() {
        return context;
    }

    @Override
    public void applyEffect(Player player, int ticks) {
        durationRemaining -= ticks;
        applyEffect(player);
    }

    abstract void applyEffect(Player player);

    @Override
    public boolean isDone() {
        return durationRemaining <= 0;
    }

    @Override
    public int getRemainingTicks() {
        return durationRemaining;
    }

    @Override
    public void setRemainingTicks(int remainingTicks) {
        durationRemaining = remainingTicks;
    }

    @Override
    public void cleanUp(Player player) {
    }
}

File: mineinabyss-features/src/main/kotlin/com/mineinabyss/curse/effects/AbstractAscensionEffect.kt
package com.mineinabyss.curse.effects

import com.github.shynixn.mccoroutine.bukkit.launch
import com.mineinabyss.components.curse.AscensionEffect
import com.mineinabyss.idofront.entities.toPlayer
import com.mineinabyss.mineinabyss.core.mineInAbyss
import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import org.bukkit.entity.Player
import kotlin.time.Duration

@Serializable
sealed class AbstractAscensionEffect : AscensionEffect {
    abstract val offset: Duration
    abstract val iterations: Int

    override fun applyEffect(player: Player, ticks: Int) {
        mineInAbyss.launch {
            delay(offset)
            repeat(iterations) {
                applyEffect(player)
                delay(duration)
            }
            cleanUp(player.uniqueId.toPlayer() ?: player) //get new player reference if player relogged
        }
    }

    abstract fun applyEffect(player: Player)

    override val isDone: Boolean
        get() = duration.inWholeMilliseconds <= 0

    override fun cleanUp(player: Player) {}
}
